### PR TITLE
Feature reload cmd

### PR DESCRIPTION
**Pull request by Jun10ng**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by Jun10ng** [comment]
@doujiang24 
@nejisama

File: pkg/stagemanager/stage_manager.go [hunk 1]
```diff
@@ -667,3 +652,62 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+// ReloadMosnProcess stops Mosn process via command line,
+// and it not support reload on windows.
+func (stm *StageManager) ReloadMosnProcess() (err error) {
+	// init
+	stm.runParamsParsedStage()
+	stm.runReconfigurationParamsInit()
+
+	mosnConfig := stm.data.config
+
+	log.StartLogger.Infof("[mosn reload] start reload mosn configurations")
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:665]:
```suggestion
	log.StartLogger.Infof("[mosn reload] start to reload mosn configurations")
```

File: pkg/stagemanager/stage_manager.go [hunk 2]
```diff
@@ -667,3 +652,62 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+	// finds process and sends SIGINT to mosn process, makes it force exit.
+	proc, p, err := getMosnProcess(mosnConfig)
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:667]: `p` is a bad name, better to use `pid` instead?

File: pkg/stagemanager/stage_manager.go [hunk 3]
```diff
@@ -667,3 +652,62 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+		log.StartLogger.Errorf("[mosn reload] fail to find process(%v), err: %v", p, err)
+		return
+	}
+
+	if err = sendSignal2Mosn(proc, p, syscall.SIGHUP); err != nil {
+		log.StartLogger.Errorf("[mosn reload] fail to send HUP signal to mosn process(%v), err: %v", p, err)
+		time.Sleep(100 * time.Millisecond) // waiting logs output
+		return
+	}
+
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:678]: we'd better wait the old process is quitted, and the new process is started, before return.

Comment (Jun10ng) [reply pkg/stagemanager/stage_manager.go:678]: ok, using signal to check process state or admin/state api?

File: pkg/stagemanager/stage_manager.go (hunk 2 above)
Comment (Jun10ng) [reply pkg/stagemanager/stage_manager.go]: pid is a package imported and I already changed the var name as `procid`

File: pkg/stagemanager/stage_manager.go (hunk 3 above)
Comment (doujiang24) [reply pkg/stagemanager/stage_manager.go:678]: checking process state should be fine.

Comment (Jun10ng) [reply pkg/stagemanager/stage_manager.go:678]: Mosn process is always alive after I execute `mosn reload`
We seem couldn't check mosn process state by sending signal?

File: pkg/stagemanager/stage_manager.go [hunk 4]
```diff
@@ -662,20 +662,74 @@ func (stm *StageManager) ReloadMosnProcess() (err error) {
...
+		if err = proc.Signal(syscall.Signal(0)); err == nil {
+			// process alive still.
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+
+		// stopped successfully
+		break
+	}
+	log.StartLogger.Infof("[mosn reload] stopped the old mosn process(%v), now waitting the new mosn process start")
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:702]: 逻辑上来说，是新进程启动之后，老进程才会退出

通常情况下，是新旧进程是会同时存在一段时间的，如果开启了平滑升级，可能共存时间会更长

在一个循环里，同时检查新旧进程，应该会更好一些

File: pkg/stagemanager/stage_manager.go [hunk 5]
```diff
@@ -662,8 +647,110 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+
+		cnt++
+		if cnt%100 == 0 { //log it per 10 second.
+			log.StartLogger.Infof("[mosn reload] the old mosn process(%v) is still existing, waiting for it quiting...", procid)
+		}
+
+		if err = proc.Signal(syscall.Signal(0)); err == nil {
+			// process alive still.
+			time.Sleep(100 * time.Millisecond)
+			continue
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:696]: 虽然是一个 for 循环了，不过还是先检查老进程是否退出呢

从输出日志来看，最好是这种的：
```
新进程 xx 开始运行了，等待老进程 xx 退出

> 老进程 xx 退出，新进程 xx 继续运行中，reload 成功
> 新进程 xx 退出，老进程 xx 继续运行中，reload 失败，老进程继续服务中
> 超时，新老进程一直都在
```

也就是新老进程，需要一起检查，这里可能会出现多种情况的
如果新进程启动失败，是会回滚到老进程运行的。

File: pkg/stagemanager/stage_manager.go [hunk 6]
```diff
@@ -662,8 +647,121 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+		if time.Now().After(t) {
+			log.StartLogger.Errorf("[mosn reload] the new process(%v) and old process(%v) are still running after waiting for %v, ignore it and quiting ...", newProcID, oldProcID, t)
+			time.Sleep(100 * time.Millisecond) // waiting logs output
+			return
+		}
+
+		// continue: new run old run
+		if newProcRun && oldProcRun { // log per 10s
+			log.StartLogger.Infof("[mosn reload] the new process(%v) and old process(%v) are running, wait the old process exit", newProcID, oldProcID)
+			time.Sleep(10 * time.Second)
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:716]: 这里 sleep 10 秒，太长了点呢，最好另外搞个变量计算日志输出

File: pkg/stagemanager/stage_manager.go [hunk 7]
```diff
@@ -662,8 +647,121 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+	}
+
+	if err = sendSignal2Mosn(oldProc, oldProcID, syscall.SIGHUP); err != nil {
+		log.StartLogger.Errorf("[mosn reload] fail to send HUP signal to mosn process(%v), err: %v", oldProcID, err)
+		time.Sleep(100 * time.Millisecond) // waiting logs output
+		return
+	}
+
+	// wait pid file re-generate
+	time.Sleep(3 * time.Second)
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:681]: 这里固定 sleep 3 秒，不是太好，最好可以挪到 for 循环里
通过 oldProcId == newProcId 来判断是否新进程起来了

Comment (Jun10ng) [reply pkg/stagemanager/stage_manager.go]: 我把获取新进程的代码也移进去了，这样看起来更流畅点

File: pkg/stagemanager/stage_manager.go (hunk 6 above)
Comment (Jun10ng) [reply pkg/stagemanager/stage_manager.go]: 好的～

File: pkg/stagemanager/stage_manager.go [hunk 8]
```diff
@@ -662,8 +647,124 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+
+	mosnConfig := stm.data.config
+
+	log.StartLogger.Infof("[mosn reload] start to reload mosn configurations")
+	// get current process info
+	var oldProc *os.Process
+	var oldProcID int
+	if oldProc, oldProcID, err = getMosnProcess(mosnConfig); err != nil {
+		log.StartLogger.Errorf("[mosn reload] fail to find process(%v), err: %v", oldProcID, err)
+		return
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:671]: also need to sleep for a while first.

File: pkg/stagemanager/stage_manager.go [hunk 9]
```diff
@@ -662,8 +647,124 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+		if newProc.Signal(syscall.Signal(0)) != nil {
+			newProcRun = false
+		}
+
+		// time-out force exit: both new & old processes are running while run out the time
+		if time.Now().After(t) {
+			log.StartLogger.Errorf("[mosn reload] the new process(%v) and old process(%v) are still running after waiting for %v, ignore it and quiting ...", newProcID, oldProcID, t)
+			time.Sleep(100 * time.Millisecond) // waiting logs output
+			return
+		}
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:712]: 这一段的位置感觉不太对，可以考虑挪到 for 循环的最前面（因为前面刚判断了进程是否存在，这里就直接说都存在，不太合理）
并且，把 688 行的判断逻辑，也合并到一起。

File: pkg/stagemanager/stage_manager.go [hunk 10]
```diff
@@ -662,8 +647,124 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+
+		// first time loop in: try to get new process info
+		if newProcID == 0 {
+			if newProc, newProcID, err = getMosnProcess(mosnConfig); err != nil {
+				if time.Now().After(t) { // timeout
+					log.StartLogger.Errorf("[mosn reload] fail to find new process(%v), err: %v", newProcID, err)
+					time.Sleep(100 * time.Millisecond) // waiting logs output
+					return
+				}
+				continue
```
Comment (doujiang24) [inline pkg/stagemanager/stage_manager.go:693]: 这个 continue 之前，需要 sleep 一会的

Comment (doujiang24) [reply pkg/stagemanager/stage_manager.go:693]: 还可以，在一段时间后（比如 3s）没有等到新进程，每一段时间（比如 10s）输出一次等待的 warning 日志

**Review by doujiang24 — APPROVED** [review]

File: cmd/mosn/main/control.go [hunk 11]
```diff
@@ -235,8 +235,19 @@ var (
 	cmdReload = cli.Command{
 		Name:  "reload",
 		Usage: "reconfiguration",
+		Flags: []cli.Flag{
```
Comment (nejisama) [inline cmd/mosn/main/control.go:238]: 所有main的改动，都要同步istio 目录下的main改动

File: pkg/stagemanager/stage_manager.go [hunk 12]
```diff
@@ -216,7 +216,8 @@ func (stm *StageManager) runInitStage() {
 	log.StartLogger.Infof("init stage cost: %v", time.Since(st))
 }
 
-func (stm *StageManager) runStopInit() {
+// runReconfigurationParamsInit to reload mosn configurations from cmd.
+func (stm *StageManager) runReconfigurationParamsInit() {
```
Comment (nejisama) [inline pkg/stagemanager/stage_manager.go:220]: 这里把stopInit给改名了，然后又把这个函数给StopMosnProcess调用 是不是很奇怪，这里是不是可以考虑和initStage复用起来比较合适

File: pkg/stagemanager/stage_manager.go [hunk 13]
```diff
@@ -662,8 +647,127 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+		// reload failed exit: new exit old run
+		if !newProcRun && oldProcRun {
+			log.StartLogger.Errorf("[mosn reload]reload mosn failed. The old process(%v) still running and new process exited(%v)", oldProcID, newProcID)
+			time.Sleep(100 * time.Millisecond) // waiting logs output
+			return
+		}
+	}
+}
+
+func getMosnProcess(cfg *v2.MOSNConfig) (mosnProcess *os.Process, p int, err error) {
```
Comment (nejisama) [inline pkg/stagemanager/stage_manager.go:742]: os.Process就包含PID 没有必要重复return

File: pkg/stagemanager/stage_manager.go [hunk 14]
```diff
@@ -662,8 +647,127 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+	// finds process and sends SIGINT to mosn process, makes it force exit.
+	mosnProcess, err = os.FindProcess(p)
+	if err != nil {
+		log.StartLogger.Errorf("[mosn] fail to find process(%v), err: %v", p, err)
+		return
+	}
+	return
+}
+
+func sendSignal2Mosn(proc *os.Process, p int, sig syscall.Signal) (err error) {
```
Comment (nejisama) [inline pkg/stagemanager/stage_manager.go:758]: 这也是一样的 process里的pid就可以了

File: pkg/stagemanager/stage_manager.go [hunk 15]
```diff
@@ -662,8 +647,127 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+		return
+	}
+
+	if err = sendSignal2Mosn(oldProc, oldProcID, syscall.SIGHUP); err != nil {
+		log.StartLogger.Errorf("[mosn reload] fail to send HUP signal to mosn process(%v), err: %v", oldProcID, err)
 		time.Sleep(100 * time.Millisecond) // waiting logs output
 		return
 	}
+
+	t, cnt, newProcRun, oldProcRun := time.Now().Add(360*time.Second), 0, true, true
```
Comment (nejisama) [inline pkg/stagemanager/stage_manager.go:681]: 这里360s的magic number 还是重新定义一下

File: pkg/stagemanager/stage_manager.go [hunk 16]
```diff
@@ -662,8 +647,127 @@ func (stm *StageManager) StopMosnProcess() (err error) {
...
+	newProc, newProcID := &os.Process{}, int(0)
+	for {
```
Comment (nejisama) [inline pkg/stagemanager/stage_manager.go:683]: 这里那么大一段逻辑，其实就是在发送完热升级信号以后，进行定期的检查和输出，是不是没有必要那么麻烦 不断的sleep & if， 搞一个timer 执行各种状态判断就好了

**Review by nejisama — COMMENTED** [review]
目前的reload是靠向旧进程发送HUP信号触发热升级流程执行reload，但是MOSN是可以通过配置关闭热升级的，有没有测试过这种情况下，reload是什么结果？

**Comment by Jun10ng** [comment]
> 目前的reload是靠向旧进程发送HUP信号触发热升级流程执行reload，但是MOSN是可以通过配置关闭热升级的，有没有测试过这种情况下，reload是什么结果？

如果打开热升级执行 reload的话，新进程会向老mosn进程向发热升级信号，老进程会报错
 "[server] SIGHUP received: current state expected running while got 13"，最终导致无法执行reload，这个问题我在最新的commit内修复了，执行reload的时候不发送热升级信号。

**Comment by Jun10ng** [comment]
> 补充对应的文档到MOSN官网，需要说明各种细节。 比如mosn reload -c config.json 并不是用config.json去reload，还是用原来start的配置进行reload的等

好的，等PR合进好，我再提个PR到mosn.io

File: pkg/mosn/mosn.go [hunk 17]
```diff
@@ -127,7 +128,7 @@ func (m *Mosn) inheritConfig(c *v2.MOSNConfig) (err error) {
 	server.EnableInheritOldMosnconfig(c.InheritOldMosnconfig)
 
 	// default is graceful mode, turn graceful off by set it to false
-	if !c.DisableUpgrade && server.IsReconfigure() {
+	if !m.ReloadProcess && !c.DisableUpgrade && server.IsReconfigure() {
```
Comment (doujiang24) [inline pkg/mosn/mosn.go:131]: maybe we can just reuse the `DisableUpgrade`? just set the `DisableUpgrade` flag instead.

Comment (Jun10ng) [reply pkg/mosn/mosn.go:131]: DisableUpgrade is got from the config.json file. Do you mean we just change disable_upgrade directly?

Comment (doujiang24) [reply pkg/mosn/mosn.go:131]: yes, overwrite the `DisableUpgrade` after getting `c` from `config.json`.

Comment (Jun10ng) [reply pkg/mosn/mosn.go:131]: > yes, overwrite the `DisableUpgrade` after getting `c` from `config.json`.

okay.

**Comment by Jun10ng** [comment]
"Error: -01 07:41:19,967 [ERROR] [config] network filter create failed, type:tcp_proxy, error: unsupported network filter type: tcp_proxy
Error: -01 07:41:19,967 [ERROR] [config] network filter factories len is 0, listener: &{ListenerConfig:{Name:0.0.0.0_80 Type: AddrConfig: BindToPort:false Network: UseOriginalDst:false "

这个CI 失败看起来不是我提交的代码影响的。

File: cmd/mosn/main/control.go [hunk 18]
```diff
@@ -236,8 +236,24 @@ var (
...
+			},
+		},
 		Action: func(c *cli.Context) error {
-			return nil
+			app := mosn.NewMosn()
+			stm := stagemanager.InitStageManager(c, c.String("config"), app)
+			stm.AppendInitStage(mosn.InitDefaultPath)
+			stm.AppendInitStage(func(cfg *v2.MOSNConfig) {
+				// disable upgrade option when mosn reloading
+				// avoiding trigger old mosn turn to state 13.
```
Comment (doujiang24) [inline cmd/mosn/main/control.go:253]:
```suggestion
				// avoiding trigger old mosn upgrading.
```
`state 13` is not readable.

**Comment by doujiang24** [comment]
```
istio/istio152/main/control.go:21:2: imported and not used: "fmt"
```
https://github.com/mosn/mosn/runs/8153277565?check_suite_focus=true#step:5:222
@Jun10ng 看起来是这个错误呢

**Review by doujiang24 — APPROVED** [review]
